### PR TITLE
Allow checkpoint and restore on non-deterministic expressions in GpuFilter and GpuProject

### PR DESCRIPTION
**Pull request by firestarman**
fix https://github.com/NVIDIA/spark-rapids/issues/7865

This PR is to support `checkpoint` and `restore` on non-deterministic expressions in `GpuFilter` and `GpuProject`.

- Introduce a new interface named `Retryable` in Java to supprt Java UDF in the future for end users. A non-deterministic expression can implement this interface to make it retryable.
- Introduce a new class named `RapidsXORShiftRandom` to provide access to the internal hashed seed. This is used by `GpuRand` to implement the `Retryable` interface.
- Removed the useless `CheckpointRestore` trait, replaced by the new `Retryable` interface.
- Update `GpuRand` to support `checkpoint` and `restore`.
- Introduce two new memebers in `GpuExpression`, they are `selfNonDeterministic` and `retryable`. `retryable` is used to tell whether an expression is retryable. It will cover its children. While `selfNonDeterministic` indicates whether an expression itself is non-deterministic when its "deterministic" is false, excluding its children. An expression is actually a tree, and `deterministic` being false means there is at least one tree node is non-deterministic, but we need to know the exact nodes which are non-deterministic to check if it implements the `Retryable`. So `selfNonDeterministic` is created.

TODO
- [x] add tests

### CODE REVIEW

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

File: sql-plugin/src/main/java/com/nvidia/spark/Retryable.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * the whole process runs on Spark executors.
+ *
+ * Retry framework expects the "runWorkOnGpu" always outputs the same result when running
+ * it multiple times in a retry. So if "runWorkOnGpu" is non-deterministic, it can not be
+ * used by Retry framework.
+ * The "Retryable" is designed for this kind of cases. By implementing this interface,
+ * "runWorkOnGpu" can become deterministic inside a retry process, making it usable for
+ * Retry framework to improve the stability.
+ */
+public interface Retryable {
```
Comment (revans2) [inline sql-plugin/src/main/java/com/nvidia/spark/Retryable.java:52]: Do we need to add this interface to the list of classes that should be outside of the parallel worlds class loader? I guess that can happen when we make this available to be something a UDF can implement.

**Review by revans2 — APPROVED** [review]
Really just some nits on the docs, and they are only nits because I am thinking ahead to when we hopefully allow this for GpuUdfs.

File: sql-plugin/src/main/java/com/nvidia/spark/Retryable.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark;
+
+/**
+ * An interface that can be used by Retry framework of RAPIDS Plugin to handle the GPU OOMs.
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/java/com/nvidia/spark/Retryable.java:20]: sure

File: sql-plugin/src/main/java/com/nvidia/spark/Retryable.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ * Retry framework expects the "runWorkOnGpu" always outputs the same result when running
+ * it multiple times in a retry. So if "runWorkOnGpu" is non-deterministic, it can not be
+ * used by Retry framework.
+ * The "Retryable" is designed for this kind of cases. By implementing this interface,
+ * "runWorkOnGpu" can become deterministic inside a retry process, making it usable for
+ * Retry framework to improve the stability.
+ */
+public interface Retryable {
+  /**
+   * Save the state, so it can be restored in case of an OOM Retry.
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/java/com/nvidia/spark/Retryable.java:54]: sure

File: sql-plugin/src/main/java/com/nvidia/spark/Retryable.java (hunk 1 above)
Comment (firestarman) [reply sql-plugin/src/main/java/com/nvidia/spark/Retryable.java:52]: agree. I will do it

Comment (firestarman) [reply sql-plugin/src/main/java/com/nvidia/spark/Retryable.java:37]: done

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Review by revans2 — APPROVED** [review]
Looks good

**Pull request merged**